flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierSubpartitionProducerAgent.java

```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.memory;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.NettyConnectionWriter;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.NettyPayload;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.TieredStorageNettyService;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.TieredStorageNettyServiceImpl;
+
+import javax.annotation.Nullable;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** The subpartition producer agent for the memory tier. */
+class MemoryTierSubpartitionProducerAgent {
+
+    private final int subpartitionId;
+
+    private final TieredStorageNettyService nettyService;
+
+    /**
+     * The {@link NettyConnectionWriter} is used to write buffers to the netty connection.
+     *
+     * <p>Note that this field can be null before the netty connection is established.
+     */
+    @Nullable private NettyConnectionWriter nettyConnectionWriter;
+
+    private int finishedBufferIndex;
+
+    MemoryTierSubpartitionProducerAgent(
+            int subpartitionId, TieredStorageNettyService nettyService) {
+        this.subpartitionId = subpartitionId;
+        this.nettyService = nettyService;
+    }
+
+    // ------------------------------------------------------------------------
+    //  Called by MemoryTierProducerAgent
+    // ------------------------------------------------------------------------
+
+    void connectionEstablished(NettyConnectionWriter nettyConnectionWriter) {
+        this.nettyConnectionWriter = nettyConnectionWriter;
+    }
+
+    void addFinishedBuffer(Buffer buffer) {
+        NettyPayload toAddBuffer =
+                NettyPayload.newBuffer(buffer, finishedBufferIndex, subpartitionId);
+        addFinishedBuffer(toAddBuffer);
+    }
+
+    void addSegmentId(int segmentId) {
+        NettyPayload segmentNettyPayload = NettyPayload.newSegment(segmentId);
+        addFinishedBuffer(segmentNettyPayload);
+    }
+
+    void release() {
+        if (nettyConnectionWriter != null) {
+            nettyConnectionWriter.close(null);
+        }
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    private void addFinishedBuffer(NettyPayload nettyPayload) {
+        finishedBufferIndex++;
+        checkNotNull(nettyConnectionWriter).writeBuffer(nettyPayload);
+        if (nettyConnectionWriter.numQueuedBuffers() <= 1
+                && nettyService.getClass() == TieredStorageNettyServiceImpl.class) {
+            ((TieredStorageNettyServiceImpl) nettyService)
```

Comment:
Currently, we can only transfer the type of the `nettyService`. `TieredStorageNettyServiceImpl ` implementing the new interface can not resolve this,  because we can not make the `createProducerAgent` to be dependent on an implementation of the `nettyService`.

Reply:
Make sense, maybe this refactor only can be done in the future as netty shuffle plugin relies a bit on the concept of `subpartitionView` now. 😞